Move to use JavaFX

SamCarlberg opened the pull request:
This also fixes #2, #6, #14, #19, #20, #21, and #27, supersedes #7 and #22, and adds the following features:

- Change app settings on the fly
  - Show/hide metadata
  - Change server address
  - Change server port
  - Change from server mode to client mode, and back again
- Search for keys/values/types
- Delete keys and branches with the delete key (no menu needed)
- Multiple selection of keys for easy deletion

Requires wpilibsuite/ntcore#202. Jenkins builds will fail until that PR is merged and the network tables dependencies in `build.gradle` are reverted to use the desktop jar instead of the one I'm building locally.

Currently this is a mess (see: MainWindowController) and needs a lot of cleanup (and tests, and checkstyle, and...)

Still TODO:

- [x] Support raw data
- [x] Support string arrays
- [x] Support number arrays
- [x] Support boolean arrays
- [ ] Comprehensive tests

## Screenshots

### Preferences (server mode)
![screenshot from 2017-05-29 21 35 20](https://cloud.githubusercontent.com/assets/6320992/26565287/d551765a-44b7-11e7-92a2-73d101b4bee8.png)

### Preferences (client mode)
![screenshot from 2017-05-29 21 35 30](https://cloud.githubusercontent.com/assets/6320992/26565286/d55133a2-44b7-11e7-9f28-9d1a55efc58a.png)

### Connecting to a remote server
![screenshot from 2017-05-29 21 39 18](https://cloud.githubusercontent.com/assets/6320992/26565290/d5540ac8-44b7-11e7-9306-0809963fc7dc.png)

### No server connection
![screenshot from 2017-05-29 21 39 43](https://cloud.githubusercontent.com/assets/6320992/26565291/d554f334-44b7-11e7-8f95-cb6846bec777.png)

### Running a server locally
![screenshot from 2017-05-29 21 40 59](https://cloud.githubusercontent.com/assets/6320992/26565288/d552da04-44b7-11e7-9ae8-29dab933c092.png)

### Add or edit strings and numbers
![screenshot from 2017-05-29 21 41 07](https://cloud.githubusercontent.com/assets/6320992/26565289/d553e660-44b7-11e7-8e75-521caed5e086.png)

### Add or edit booleans
![screenshot from 2017-05-29 21 41 21](https://cloud.githubusercontent.com/assets/6320992/26565293/d55aea1e-44b7-11e7-9d00-aa8c7d26485f.png)

### String / number arrays
![screenshot from 2017-05-29 21 41 43](https://cloud.githubusercontent.com/assets/6320992/26565292/d55a51e4-44b7-11e7-8389-3f2335ee4a8c.png)

### Boolean arrays
![screenshot from 2017-05-29 21 42 12](https://cloud.githubusercontent.com/assets/6320992/26565294/d55b414e-44b7-11e7-85f4-275161515371.png)

### Arrays of raw bytes
![screenshot from 2017-05-29 21 42 54](https://cloud.githubusercontent.com/assets/6320992/26565295/d55e68c4-44b7-11e7-96c9-f8c0294b1994.png)

SamCarlberg committed:
Add initial JavaFX

- able to search keys/values/types
- start in client or server mode (remembers IPs and ports)
- able to make entries persistent or not
- supports String, Number, and Boolean types

SamCarlberg committed:
gitignore ntcore persistence files

SamCarlberg committed:
Fix deletes and persistence

ntcore is really inconsistent. sometimes it wants a leading slash, sometimes it doesn't

SamCarlberg committed:
Make window not resizable when showing opening prefs

SamCarlberg committed:
Use ncore JNI instead of NetworkTable abstraction

SamCarlberg committed:
Allow selection and deletion of multiple entries at once

And with the delete button, too! No need to have to use menus for it

SamCarlberg committed:
Move global preferences to its own class

SamCarlberg committed:
Move preferences to its own dialog

SamCarlberg committed:
Add app preferences pane to change settings live

Currently allows showing/hiding meatdata, hopefully in the fuure we can change IP, switch from server to client mode, etc

SamCarlberg committed:
Fix metadata displaying when it shouldn't

SamCarlberg committed:
Remove useless edit and help menus

SamCarlberg committed:
Make metadata actually show its type

SamCarlberg committed:
Make metadata regex a little more explicit

SamCarlberg committed:
Revert "Make metadata regex a little more explicit"

This reverts commit 62e3b82a89e96f43d16763e85eead91e4f55679d.

SamCarlberg committed:
Add documentation to prefs

SamCarlberg committed:
Add title to preferences dialog

SamCarlberg committed:
Make app aware of network mode

Consolidate all preferences in one dialog used at startup and during use
Be able to change between client and server mode
Hide/show metadata at will
Change server port and address (client mode only) - again, at will
Currently requires ntcore linux and wpilibsuite/ntcore#202

SamCarlberg committed:
Server ID tooltip

SamCarlberg committed:
Make entries more OO

SamCarlberg committed:
Add support for viewing arrays

Fix an NPE with initial entry values
Make raw bytes start with 0x

SamCarlberg committed:
Docs to entry classes

SamCarlberg committed:
Refactor most of the tree stuff out of MainWindowController

SamCarlberg committed:
Simplify metadata visibility code

SamCarlberg committed:
Move connection indicator to its own view and controller

SamCarlberg committed:
Show integer numbers without decimal points

SamCarlberg committed:
Clean up MainWindowController

SamCarlberg committed:
Make NetworkTableTree work when filter is set before root

Being proactive here. This issue will probably never happen anyway but it's good to have proper code just in case

SamCarlberg committed:
Add dialogs for adding arrays

SamCarlberg committed:
Add editing of string, number, boolean arrays

SamCarlberg committed:
Add support for adding/editing raw byte arrays

SamCarlberg committed:
Add separator between add arrays and add raw data menus

SamCarlberg committed:
Add checkstyle

SamCarlberg committed:
Fix preferences not using port field when in client mode

SamCarlberg committed:
Removed unused Predicates class

SamCarlberg committed:
Add tests for entry classes and preferences controller

SamCarlberg committed:
Add tests for NetworkTableUtils, ToggleSwitchListCell

SamCarlberg committed:
Add initial tests for the add entry dialogs

SamCarlberg committed:
Allow lowercase in hex values for raw data

ThadHouse commented:
Where does setting the remote name happen? Also that only works before a connection starts.

SamCarlberg commented:
There's a bit going on under the hood. If the server ID is a number, it assumes it's a team number and attempts to connect to `roborio-TEAM-frc.local`. Otherwise it's assumed to be a URL or IP address and passes that directly to ntcore.

Changing modes between client and server shuts down the current instance of ntcore and spins up a new one. Same for changing port or server ID.

ThadHouse commented:
I meant remote name. Since you tagged #14. That option should be added to that window as well.

ThadHouse commented:
Also if possible we should support arrays of IP addresses for connecting to, since thats supported by ntcore as well.

AustinShalit commented on src/main/java/edu/wpi/first/tableviewer/PreferencesController.java:
```diff
@@ -0,0 +1,84 @@
+    if (url.matches("[0-9]+")) {
+      // treat as a team number
+      url = "roborio-" + url + "-frc.local";
+    }
```
Use nt set team here instead. That way it will try all the possible ips

SamCarlberg replied:
I forgot ntcore had that. Thanks!

AustinShalit replied:
Also, you can change the regex to limit the amount of numbers to be less than 4

SamCarlberg replied:
I think `"[0-9]{1,5}"` would be fine. 0-5 digits is enough for a while

AustinShalit commented:
@thadhouse Are you thinking a comma separated list for IP addresses?

ThadHouse commented:
@AustinShalit sure they could work. I know we were discussing parallelizing connections in ntcore, so adding more isn't an issue.

SamCarlberg commented:
I feel like multiple connections is out of scope for this project

ThadHouse commented:
True. We can add that at a later date. Along with remote name if thats too much.

SamCarlberg committed:
Fix some networking issues. Use NetworkTable.setTeam when approriate

Log changes to preferences

SamCarlberg committed:
Allow renaming of entries. Use ntcore JNI for everything

Implemented as a delete, then put

SamCarlberg committed:
Checkstyle

SamCarlberg committed:
Add styles for entry dialogs, improve switches for boolean arrays

SamCarlberg committed:
Separate NetworkTableTree from ntcore

SamCarlberg committed:
Remove empty parent tables when an entry is deleted

SamCarlberg committed:
Add tests for components package

SamCarlberg committed:
100% test coverage on PreferencesController

SamCarlberg committed:
Add FxHelper for running and waiting for stuff on the application thread

SamCarlberg committed:
Add tests for AddBytesDialog

SamCarlberg committed:
Add tests for MainWindowController, ConnectionIndicatorController

Also add helpful superclass for UI tests to automatically clean up windows after each test

SamCarlberg committed:
Use ntcore desktop after merge of wpilibsuite/ntcore#202

SamCarlberg committed:
Add -Pheadless gradle flag for headless tests

SamCarlberg committed:
Run UI tests on travis in headless mode

SamCarlberg commented:
@JLLeitschuh Any idea why the travis UI tests are failing? Looks like there's no JavaFX installed in the image

AustinShalit commented:
Try this: https://github.com/wpilibsuite/OutlineViewer/pull/24/files#diff-354f30a63fb0907d4ad57269548329e3

AustinShalit committed:
Fix Travis & Jenkins

AustinShalit committed:
Remove context menu tests

AustinShalit commented on src/main/java/edu/wpi/first/tableviewer/Main.java:
```diff
@@ -39,6 +39,8 @@ public void start(Stage primaryStage) throws IOException {
     primaryStage.setScene(new Scene(mainWindow));
     primaryStage.show();
+    primaryStage.setMinWidth(600);
+    primaryStage.setMinHeight(450);
```
Should this be in FXML?

AustinShalit committed:
Remove broken tests & update build
